pkg/awsutils/awsutils.go

```diff
@@ -1515,3 +1520,25 @@ func (cache *EC2InstanceMetadataCache) IsUnmanagedENI(eniID string) bool {
 	}
 	return false
 }
+
+func (cache *EC2InstanceMetadataCache) getENIsFromPaginatedDescribeNetworkInterfaces(
+	input *ec2.DescribeNetworkInterfacesInput) ([]*ec2.NetworkInterface, error) {
+	outputENIs := make([]*ec2.NetworkInterface, 0)
+	pageNum := 0
+	log.Debugf("Paginating describe ENI has page size: %d", *input.MaxResults)
+	pageFn := func(output *ec2.DescribeNetworkInterfacesOutput, lastPage bool) (nextPage bool) {
+		pageNum++
+		log.Infof("EC2 DescribeNetworkInterfaces succeeded with %d results on page %d",
+			len(output.NetworkInterfaces), pageNum)
+
+		outputENIs = append(outputENIs, output.NetworkInterfaces...)
+		// if no more ENI, we want to exit even it is not the last page to avoid many blank calls.
+		if len(output.NetworkInterfaces) == 0 {
+			lastPage = true
+		}
+		return !lastPage
```

Comment:
You can simplify this by just saying 

return output.NextToken != nil

Reply:
If possible we should filter out ENIs which are not required like it should have description aws-k8s etc. @jayanthvn looking at https://github.com/aws/amazon-vpc-cni-k8s/blob/508c28dc04e11393c6cdb68bfba0e8f1da07b3a0/pkg/awsutils/awsutils.go#L1410 seems like ipamd will clean up leaked ENIs for other clusters in the same account as well.

Reply:
The cleanup does look for all ENIs in the same account.

Reply:
@SaranBalaji90 Yes cleanup fetches all "available" ENIs with aws-K8S-* description in the cx account. This PR https://github.com/aws/amazon-vpc-cni-k8s/pull/1109 mitigates that issue for some extent. But yes cluster tag would surely restrict cleanup within the node's cluster. To use clusterName we need to have the CLUSTER_NAME env variable set or fetch from ec2 instance tag key -  "kubernetes.io/cluster/jay-v17-cluster = owned ".

Reply:
Fwiw, you can also just always `return true` if you want to process every page, and then stop.  There's no need to use `lastPage` ([`request.Pagination`](https://docs.aws.amazon.com/sdk-for-go/api/aws/request/#Pagination) uses the same `p.HasNextPage()` check for both the `lastPage` variable, and when to stop the `p.Next()` iterator)

Reply:
> Yes cleanup fetches all "available" ENIs with aws-K8S-* description in the cx account. 

(Aside: It's a larger change, but fwiw we could also run this loop _once_ per cluster, instead of on every node.  Just move the code out into a separate "cleanup" pod that we run as a Deployment replicas=1.  It doesn't make much sense to do a global cleanup _times_ the number of nodes :worried: )

Reply:
An alternate could be to do it under leader election from one of the node and don't do it if it has already happened in the last x interval.

Reply:
> An alternate could be to do it under leader election from one of the node and don't do it if it has already happened in the last x interval.

￼Agreed, using a Lease is the other obvious solution.  I think this means ~10k watches (ie: 1/node) from every aws-node pod on a shared Lease/ConfigMap, which is ok - and better than running the loop 10k times as we currently do.